stats/opentelemetry/internal/grpc_trace_bin_propagator_test.go

```diff
@@ -0,0 +1,252 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package internal
+
+import (
+	"context"
+	"encoding/base64"
+	"reflect"
+	"testing"
+
+	otelpropagation "go.opentelemetry.io/otel/propagation"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+	itracing "google.golang.org/grpc/stats/opentelemetry/internal/tracing"
+)
+
+// TODO: Move out of internal as part of open telemetry API
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+// TestInject_FastPath verifies that the GRPCTraceBinPropagator correctly
+// injects OpenTelemetry trace context data using the CustomCarrier.
+//
+// It is called the fast path because it injects the trace context directly in
+// binary format.
+func (s) TestInject_FastPath(t *testing.T) {
+	p := GRPCTraceBinPropagator{}
+	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
+		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
+		TraceFlags: oteltrace.FlagsSampled,
+	})
+	tCtx, tCancel := context.WithCancel(context.Background())
+	tCtx = oteltrace.ContextWithSpanContext(tCtx, sc)
+	defer tCancel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	c := itracing.NewCustomCarrier(metadata.NewOutgoingContext(ctx, metadata.MD{}))
+	p.Inject(tCtx, c)
+
+	got := stats.OutgoingTrace(c.Context())
+	want := binary(sc)
+	if string(got) != string(want) {
+		t.Fatalf("got = %v, want %v", got, want)
+	}
+}
+
+// TestInject_SlowPath verifies that the GRPCTraceBinPropagator correctly
+// injects OpenTelemetry trace context data using any other text based carrier.
+//
+// It is called the slow path because it base64 encodes the binary trace
+// context before injecting it.
+func (s) TestInject_SlowPath(t *testing.T) {
+	p := GRPCTraceBinPropagator{}
+	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
+		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
+		TraceFlags: oteltrace.FlagsSampled,
+	})
+	tCtx, tCancel := context.WithCancel(context.Background())
+	tCtx = oteltrace.ContextWithSpanContext(tCtx, sc)
+	defer tCancel()
+
+	c := otelpropagation.MapCarrier{}
+	p.Inject(tCtx, c)
+
+	got := c.Get(itracing.GRPCTraceBinHeaderKey)
+	want := base64.StdEncoding.EncodeToString(binary(sc))
+	if got != want {
+		t.Fatalf("got = %v, want %v", got, want)
+	}
+}
```

Comment:
Same here, let's try and make this a t-test (or if too many distinct operations, merge one of them with a t-test with errors as described in next sentence). And then for the merged t-test add tests for the two cases that return nil (invalid span context and nil binary data).